src/lib/ppx_register_event/register_event.ml

```diff
@@ -0,0 +1,201 @@
+  let ctx0 = init () in
+  let ctx1 = feed_string ctx0 name in
+  get ctx1 |> to_raw_string
+
+(* we omit an explicit `Core_kernel`, because the Ppx_deriving_yojson functions
+   generate uses of `Result` without it, so that module will need to be opened in
+   any case
+*)
+let core_kernel_result f = String.concat ["Result"; f] ~sep:"."
+
+let type_ext_str ~options ~path (ty_ext : type_extension) =
+  let add_module_qualifier s = String.concat ~sep:"." [module_name; s] in
+  let ctor, ctor_args, label_decls =
+    match ty_ext.ptyext_constructors with
+    (* record argument *)
+    | [ { pext_name
+        ; pext_kind= Pext_decl ((Pcstr_record labels as record), None)
+        ; _ } ] ->
+        (pext_name.txt, record, labels)
+    (* no arguments *)
+    | [{pext_name; pext_kind= Pext_decl ((Pcstr_tuple [] as tuple), None); _}]
+      ->
+        (pext_name.txt, tuple, [])
+    | _ ->
+        Location.raise_errorf ~loc:ty_ext.ptyext_path.loc
+          "Constructor in type extension must take a single record argument, \
+           or no argument"
+  in
+  let label_names =
+    List.map label_decls ~f:(fun {pld_name= {txt; _}; _} -> txt)
+  in
+  let has_record_arg = not @@ List.is_empty label_names in
+  let msg, msg_loc =
+    match List.Assoc.find options "msg" ~equal:String.equal with
+    | Some expr -> (
+      match expr.pexp_desc with
+      | Pexp_constant (Pconst_string (s, _)) ->
+          (s, expr.pexp_loc)
+      | _ ->
+          Location.raise_errorf ~loc:expr.pexp_loc
+            "The msg option must be a string constant" )
+    | None ->
+        if has_record_arg then
+          let fields =
+            List.map label_names ~f:(fun label -> sprintf "%s=$%s" label label)
+          in
+          ( sprintf "%s {%s}" ctor (String.concat ~sep:"; " fields)
+          , Location.none )
+        else (sprintf "%s" ctor, Location.none)
+  in
+  (* check that every interpolation point $foo in msg has a matching label;
+     OK to have extra labels not mentioned in message
+  *)
+  let len = String.length msg in
+  let interpolates =
+    let re = Str.regexp "\\$[a-z]\\([a-z]\\|[0-9]\\)*" in
+    let rec loop start acc =
+      if start >= len then acc
+      else
+        try
+          let offs = Str.search_forward re msg start in
+          let s = Str.matched_string msg in
+          let s_len = String.length s in
+          loop (offs + s_len) (String.sub s ~pos:1 ~len:(s_len - 1) :: acc)
+        with _ ->
+          (* wild-card match; actual match is deprecated Not_found *)
+          acc
+    in
+    loop 0 []
+  in
+  List.iter interpolates ~f:(fun interp ->
+      if not (List.mem label_names interp ~equal:String.equal) then
+        Location.raise_errorf ~loc:msg_loc
+          (Scanf.format_from_string
+             (sprintf
+                "The msg contains interpolation point \"$%s\" which is not a \
+                 field in the record"
+                interp)
+             "") ) ;
```

Comment:
Nit: Lift this to be its own function; it's orthogonal to all the other logic here.